lib/semantics/tools.h

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef FORTRAN_SEMANTICS_TOOLS_H_
+#define FORTRAN_SEMANTICS_TOOLS_H_
+
+// Simple predicates and look-up functions that are best defined
+// canonically for use in semantic checking.
+
+#include "scope.h"
+#include "symbol.h"
+#include "type.h"
+#include "../evaluate/variable.h"
+
+namespace Fortran::semantics {
+
+const Symbol *FindCommonBlockContaining(const Symbol &object);
+const Scope *FindProgramUnitContaining(const Scope &);
+const Scope *FindProgramUnitContaining(const Symbol &);
+const Scope *FindPureFunctionContaining(const Scope *);
+
+bool IsCommonBlockContaining(const Symbol &block, const Symbol &object);
+bool IsAncestor(const Scope *maybeAncestor, const Scope &maybeDescendent);
```

Comment:
`IsAncestor` could be referring to submodules and modules. So I think this needs a different name (or at least a comment). Maybe `IsAncestorScope` or `IsNestedIn` with the arguments reversed.

Reply:
Ok, done.